tests/tests/test_docker_compose.py

```diff
@@ -0,0 +1,218 @@
+    def make_artifact(filename, artifact_name):
+        cmd = [
+            "gen_docker-compose",
+            "--artifact-name",
+            artifact_name,
+            "--device-type",
+            "qemux86-64",
+            "--output-path",
+            filename,
+            "--manifests-dir",
+            manifests_dir,
+            "--images-dir",
+            images_dir,
+            "--project-name",
+            project_name,
+        ]
+        subprocess.check_call(cmd)
+        return filename
+
+    return make_artifact
+
+
+class TestDockerCompose(MenderTesting):
+    def test_successful_rollback(self, standard_setup_virtualization):
+        env = standard_setup_virtualization
+        mender_device = env.device
+
+        devauth = DeviceAuthV2(env.auth)
+        deploy = Deployments(env.auth, devauth)
+
+        devices = devauth.get_devices_status("accepted")
+        assert len(devices) == 1
+        device_id = devices[0]["id"]
+
+        services = [
+            DockerService(name="test1", image="test-container-image1"),
+            DockerService(name="test2", image="test-container-image2"),
+        ]
+
+        with create_test_manifest(services) as (manifests_dir, images_dir):
+            deployment_id, _ = common_update_procedure(
+                verify_status=True,
+                devices=[device_id],
+                make_artifact=make_docker_compose_artifact(
+                    manifests_dir, "test", images_dir
+                ),
+                devauth=devauth,
+                deploy=deploy,
+            )
+
+        deploy.check_expected_status("finished", deployment_id)
+        deploy.check_expected_statistics(deployment_id, "success", 1)
+
+        docker_ps = mender_device.run("docker ps")
+        assert "test-container-image1" in docker_ps
+        assert "test-container-image2" in docker_ps
+
+        # Trigger a rollback by providing a non-existing image
+        services = [
+            DockerService(name="test1", image="non-existing-image", build_image=False),
+            DockerService(name="test2", image="test-container-image3"),
+        ]
+
+        with create_test_manifest(services) as (manifests_dir, images_dir):
+            deployment_id, _ = common_update_procedure(
+                verify_status=True,
+                devices=[device_id],
+                make_artifact=make_docker_compose_artifact(
+                    manifests_dir, "test", images_dir
+                ),
+                devauth=devauth,
+                deploy=deploy,
+            )
+
+        deploy.check_expected_status("finished", deployment_id)
+        deploy.check_expected_statistics(deployment_id, "failure", 1)
+
+        docker_ps = mender_device.run("docker ps")
+        assert "test-container-image1" in docker_ps
+        assert "test-container-image2" in docker_ps
+        assert not "non-existing-image" in docker_ps
+        assert not "test-container-image3" in docker_ps
```

Comment:
I was tempted to ask for a refactor of the "docker ps"  + check output. But I think this is a good and explicit way of making the tests 👌 

One improvement: please log the `docker_ps` as info into the test. This will soon or later need debugging and it will be nice to have all the "docker ps" outputs, including the passing ones before the tentatively failing assert.